### DISCUSSION TITLE
NextJS13: window.location.reload() on client component - crashing build on vercel

### DISCUSSION THREAD

**Discussion by hodeyp** [vercel/next.js#47530] [Help]
### Summary

I have a component that signs a user out of an app but upon sign out there is a requirement to reload the page. This is due to a third party library (reactfire) not clearing caches.

The client component has 'use client' at the top, and the component is a simple button, code sample attached for your reference.

As soon as I embed the component onto a server side page (app/admin) vercel throws an error when building...

```
Error occurred prerendering page "/admin". Read more: https://nextjs.org/docs/messages/prerender-error
--
18:02:25.506 | ReferenceError: location is not defined
```

I can serve and build the app locally using next dev and next build. But as soon as I try to deploy there is an issue. 

I thought the whole idea of NextJS13 was that you could have a server rendered page and then child components that are rendered client side? Am I missing something here? 

### Additional information

```js
'use client';

import { useAuth } from 'reactfire';
import Button from '@ui/button';

export default function SignOut() {

    const auth = useAuth();

    async function handleLogout() {

        await auth.signOut();

        // To clear the firestore cache reload the page
        // https://github.com/FirebaseExtended/reactfire/issues/485
        if (window && window.location) window.location.reload();

    }

    return (
        <>

            {/* <pre>{JSON.stringify(authData, null, 4)}</pre> */}

            <Button onClick={handleLogout} size={'lg'} fullWidth>
                Sign Out
            </Button>

        </>
    );

}
```
```


### Example

_No response_

**Comment by icyJoseph** [comment]
> I thought the whole idea of NextJS13 was that you could have a server rendered page and then child components that are rendered client side? Am I missing something here?

Your code is fine though, at least at first sight, but that assumption is incorrect.

Client Components are the components we have used up to this point. Imagine a bot arrives to your site to crawl it, but you don't render Client components, because it is SSR, then you'd miss out on SEO, what's even worse, shipping the app to Non JS running environments, or plain simple slow cpu/connections, would show holes in your app where the client bits should be. This might lead to other issues as well... so client components, at least in the Next.js interpretation of RSC, are rendered during SSR as well.

So you cannot do `window`, `document`, or in general DOM access without proper guards. With that out of the way....

The thing is that your code seems to have the guards in place. You window access is done within a handler, which is not executed until the user interacts with a button. 

In short, this should work. The thing that makes matters worse is that, if the handler was at all being accidentally called, I'd expect an undefined reference to `window`, because the proper way to check for presence of globals is using `typeof`, that is `typeof window !== 'undefined'`, but your code crashed on `window.location`... 

- Could it be that the problem is else where? Somewhere window is being accessed during SSR?
   - Ugly debugging technique, but if you were to place a log before AND after `await auth.signOut()`, do you see any during SSR? You should see none though.
- Could it be that somehow window is a thing in your project? even server side? maybe a third party dependency is messing with it?

**Reply by hodeyp** [reply]
Thanks @icyJoseph - what you are saying makes complete sense and is sort of what I was thinking as well. I've added debug logging to see if I can identify anywhere else that could be the root cause.

What is odd is that the code compiles and runs perfectly locally. I have also built and deployed to a container running on Google Cloud Run and this again is running perfectly. But if I try to build the code using the vercel deploy system it fails with the error message about location not being defined.

**Reply by icyJoseph** [reply]
Odd... 

- This is all within appDir right?
- Did you manage to log anything from the Vercel deployment?
- I wonder if the page is being pre-rendered for the edge runtime?
